### PR TITLE
Workflow deployments to dev with Terraform

### PR DESCRIPTION
**Pull request by tonyto**
### Context

Deploying the latest merge from `main` to a development environment using Terraform

### Changes proposed in this pull request

- builds the latest docker image
- pushes docker image to dockerhub
- terraform setups the dev environment and starts the docker image

### CODE REVIEW

File: terraform/app/modules/paas/versions.tf [hunk 1]
```diff
@@ -0,0 +1,9 @@
+terraform {
+  required_version = "~> 0.14.5"
+  required_providers {
```
Comment (guy-roberts) [inline terraform/app/modules/paas/versions.tf:3]: Comment really, but will this version number need to be updated
 over the years ?

File: terraform/app/terraform.tf [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+Store infrastructure state in a remote store (instead of local machine):
+https://www.terraform.io/docs/state/purpose.html
+*/
+# terraform {
+
+#   backend "s3" {
+#     key     = "terraform.tfstate"
+#     region  = "eu-west-2"
+#     encrypt = "true"
+#     bucket    = "eyfs-terraform-backend"
```
Comment (guy-roberts) [inline terraform/app/terraform.tf:29]: Probs put the bucket name in a secret

File: terraform/workspace-variables/dev.tfvars [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# Platform
+environment = "dev"
+app_environment = "dev"
+
+# Gov.UK PaaS
+paas_api_url = "https://api.london.cloud.service.gov.uk"
```
Comment (guy-roberts) [inline terraform/workspace-variables/dev.tfvars:6]: Into a secret ? Will it always have this value

File: terraform/workspace-variables/dev_app_env.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+---
+RACK_ENV: production
+RAILS_ENV: development
+NODE_ENV: production
+RAILS_LOG_TO_STDOUT: true
+RAILS_MAX_THREADS: 5
+RAILS_SERVE_STATIC_FILES: true
+SERVICE_TYPE: web
+DATABASE_URL: postgres://postgres@localhost:5432
```
Comment (guy-roberts) [inline terraform/workspace-variables/dev_app_env.yml:9]: Looks odd to me to have the PG port in the code

And the fact that its on the same host.  In heroku the db credentials are all in a secret

**Review by guy-roberts — APPROVED** [review]
I'll leave it up to you wether you want to make these changes.

Because it's mostly declarations of params for a process that I don't know about, I have not got many comments.

The theme is really, to make it easy for somebody to reconfigure, without learning all about Terraform.

And some ports and URLs are in the repo,  suggest putting them in secrets

It probably deserves an explanation in the README for people considering repeating the process.

File: terraform/app/terraform.tf (hunk 2 above)
Comment (tonyto) [reply terraform/app/terraform.tf:29]: This is not required for now - I'll remove

File: terraform/app/modules/paas/versions.tf (hunk 1 above)
Comment (tonyto) [reply terraform/app/modules/paas/versions.tf:3]: This is more to lock the version to not introduce breaking changes.

If it's required to run on another version of Terraform, this should be updated

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -1,36 +1,79 @@
...
+          tags: ${{ env.DOCKER_IMAGE }}
+          build-args: COMMIT_SHA=${{ github.sha }}
+
+      - name: Docker image digest
+        run: echo ${{ steps.docker_build.outputs.digest }}
+
+      - name: Deploy to dev
+        if: ${{ github.event.inputs.environment == 'dev' }}
+        env:
+          TF_VAR_paas_user: ${{ secrets.GOVPAAS_USERNAME }}
```
[the remark this answers is not in the document]
Comment (tonyto) [reply .github/workflows/deploy.yml:73]: Will create a `deployment.md` once everything is working

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -1,36 +1,79 @@
...
-        cf7 auth
-        cf7 target -o $CF_ORG -s $CF_SPACE
-        cf7 push
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+
+      - name: Login to DockerHub
+        uses: docker/login-action@v1
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
```
[the remark this answers is not in the document]
Comment (tonyto) [reply .github/workflows/deploy.yml:49]: Will create a deployment.md once everything is working

File: terraform/workspace-variables/dev.tfvars (hunk 3 above)
Comment (tonyto) [reply terraform/workspace-variables/dev.tfvars:6]: Lots will break if this changed 
https://github.com/search?q=https%3A%2F%2Fapi.london.cloud.service.gov.uk%2F&type=code

Comment (guy-roberts) [reply terraform/workspace-variables/dev.tfvars:6]: Was just thinking of hiding it from the Russians 

They know it already I suppose

**Pull request merged**